core/devmode/src/test/java/io/quarkus/dev/CompilerFlagsTest.java

```diff
@@ -0,0 +1,87 @@
+package io.quarkus.dev;
+
+import static org.junit.jupiter.api.Assertions.assertAll;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.junit.jupiter.api.Test;
+
+public class CompilerFlagsTest {
+
+    @Test
+    void nullHandling() {
+        assertAll(
+                () -> assertEquals(
+                        new CompilerFlags(null, null, null, null),
+                        new CompilerFlags(setOf(), listOf(), null, null)));
+    }
+
+    @Test
+    void defaulting() {
+        assertAll(
+                () -> assertEquals(
+                        new CompilerFlags(setOf("-a", "-b"), listOf(), null, null),
+                        new CompilerFlags(setOf(), listOf("-a", "-b"), null, null)),
+                () -> assertEquals(
+                        new CompilerFlags(setOf("-c", "-d"), listOf("-a", "-b"), null, null),
+                        new CompilerFlags(setOf(), listOf("-a", "-b", "-c", "-d"), null, null)));
+    }
+
+    @Test
+    void redundancyReduction() {
+        assertAll(
+                () -> assertEquals(
+                        new CompilerFlags(setOf("-a", "-b"), listOf(), null, null),
+                        new CompilerFlags(setOf(), listOf("-a", "-b"), null, null)),
+                () -> assertEquals(
+                        new CompilerFlags(setOf("-a", "-b", "-c"), listOf("-a", "-b"), null, null),
+                        new CompilerFlags(setOf("-c"), listOf("-a", "-b"), null, null)));
+    }
+
+    @Test
+    void sourceAndTarget() {
```

Comment:
What would happen if `--source` is set but the source flag is also set in maven's `compilerArgs`?

Reply:
The way it is implemented right now, it would keep both.

I think there are two reasonable choices:

1. prefer the explicit compiler flags and ignore `source`/`target` if they conflict (rationale: if you are setting compiler flags explicitly, you probably expect them to be applied; the `source` and `target` settings might just have been taken from their defaults and you may not even be aware of them)
 2. pass both

I wonder what the Maven compiler plugin does here.  It would probably be best to try and be consistent with it.  My guess is (2), but I'll see if I can find out more so I can say with more certainty.

We should definitely add a unit test that at least documents the behavior in either case, though.

Reply:
OK, looking at `org.codehaus.plexus.compiler.javac.JavacCompiler` (`plexus-compiler-javac-2.8.5-sources.jar`) it does look like it simply keeps both.  It adds any custom flags _after_ the `-source` and `-target` settings derived from the configuration, so we should probably make sure that we do that, too.

Reply:
Makes sense

Reply:
Done.